src/storage/v2/replication/replication_client.cpp

```diff
@@ -41,12 +41,50 @@ Storage::ReplicationClient::ReplicationClient(std::string name, Storage *storage
   }
 
   rpc_client_.emplace(endpoint, &*rpc_context_);
-  TryInitializeClient();
+  TryInitializeClientSync();
 
   if (config.timeout && replica_state_ != replication::ReplicaState::INVALID) {
     timeout_.emplace(*config.timeout);
     timeout_dispatcher_.emplace();
   }
+
+  // Help the user to get the most accurate replica state possible.
+  if (config.replica_check_delay > 0) {
+    replica_checker_.Run("Replica Checker", std::chrono::duration<uint64_t>(config.replica_check_delay),
+                         [&] { FrequentCheck(); });
+  }
+}
+
+void Storage::ReplicationClient::TryInitializeClientAsync() {
+  thread_pool_.AddTask([this] {
+    rpc_client_->Abort();
+    this->TryInitializeClientSync();
+  });
+}
+
+void Storage::ReplicationClient::FrequentCheck() {
+  bool is_success = true;
+  try {
+    auto stream{rpc_client_->Stream<replication::FrequentHeartbeatRpc>()};
+    const auto response = stream.AwaitResponse();
+    if (!response.success) {
+      is_success = false;
+    }
+  } catch (const rpc::RpcFailedException &) {
+    is_success = false;
+  }
+  // States: READY, REPLICATING, RECOVERY, INVALID
+  // If success && ready, replicating, recovery -> stay the same because something good is going on.
+  // If success && INVALID -> [it's possible that replica came back to life] -> TryInitializeClient.
+  // If fail -> [replica is not reachable at all] -> INVALID state.
+  // NOTE: TryInitializeClient might return nothing if there is a branching point.
+  if (is_success && replica_state_.load() != replication::ReplicaState::INVALID) {
+    // Pass because all seems fine, replica is in some valid state.
+  } else if (is_success && replica_state_.load() == replication::ReplicaState::INVALID) {
+    TryInitializeClientAsync();
+  } else if (!is_success) {
+    replica_state_.store(replication::ReplicaState::INVALID);
+  }
```

Comment:
```suggestion
  if (!is_success) {
    replica_state_.store(replication::ReplicaState::INVALID);
    return;
  }
  if (replica_state_.load() == replication::ReplicaState::INVALID) {
    TryInitializeClientAsync();
  }
```

Reply:
Yep, it should be the same but simplified 😄